File: src/main/java/com/snackgame/server/member/domain/StatusService.java
package com.snackgame.server.member.domain;

import org.springframework.stereotype.Service;

@Service
public class StatusService {

    public Status updateStatus(Status status, double score) {
        double remainExp = status.getExp() + score;
        double expRequireForLevel = calculateExpByLevel(status.getLevel());

        while (isExpExceeded(remainExp, expRequireForLevel)) {
            levelUp(status);
            remainExp -= expRequireForLevel;
            expRequireForLevel = calculateExpByLevel(status.getLevel());
        }

        status.changeExpTo(remainExp);
        return status;
    }

    private void levelUp(Status status) {
        status.addLevel(1);
    }

    private boolean isExpExceeded(double afterExp, double beforeExp) {
        return afterExp > beforeExp;
    }

    private double calculateExpByLevel(long level) {
        return Math.pow(1.2, level) * 200;
    }

}

File: src/main/java/com/snackgame/server/status/StatusService.kt
package com.snackgame.server.status

import com.snackgame.server.game.metadata.Metadata
import com.snackgame.server.game.session.event.SessionEndEvent
import com.snackgame.server.member.domain.MemberRepository
import com.snackgame.server.member.domain.getBy
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.transaction.event.TransactionPhase
import org.springframework.transaction.event.TransactionalEventListener

@Service
class StatusService(private val memberRepository: MemberRepository) {

    @TransactionalEventListener(phase = TransactionPhase.BEFORE_COMMIT)
    @Transactional
    fun onSessionEnd(session: SessionEndEvent) {
        when (session.metadata) {
            Metadata.SNACK_GAME,
            Metadata.APPLE_GAME,
            Metadata.SNACK_GAME_INFINITE,
            Metadata.SNACK_GAME_BIZ -> addExpWithScore(session)
        }
    }

    private fun addExpWithScore(session: SessionEndEvent) {
        val member = memberRepository.getBy(session.ownerId)
        member.status.addExp(session.score.toDouble())
    }
}
